File: Python/BinarySearch/Binary Search in forest.py
//https://www.geeksforgeeks.org/problems/binary-search-in-forest--141631/1
class Solution:
    def checkheight(self, tree, n, h):
        wood = 0
        for i in range(n):
            if tree[i] > h:
                wood += tree[i] - h
        return wood

    def find_height(self, tree, n, k):
        low, high = 0, max(tree)
        
        while low <= high:
            mid = (low + high) // 2
            wood = self.checkheight(tree, n, mid)
            if wood == k:
                return mid
            elif wood < k:
                high = mid - 1
            else:
                low = mid + 1
                

File: Java/BinarySearch/Binary Search in forest.java
//https://www.geeksforgeeks.org/problems/binary-search-in-forest--141631/1
class Solution
{
    static int checkheight(int[] tree, int n, int h) {
        int wood = 0;
        for (int i = 0; i < n; i++) {
            if (tree[i] > h) {
                wood += tree[i] - h;
            }
        }
        return wood;
    }

    static int find_height(int[] tree, int n, int k) {
        int low = 0, high = tree[0];
        for (int i = 0; i < n; i++) {
            high = Math.max(high, tree[i]);
        }

        while (low <= high) {
            int mid = (low + high) / 2;
            int wood = checkheight(tree, n, mid);
            if (wood == k) {
                return mid;
            } else if (wood < k) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
}
